### PR TITLE
rangefeed: add rangefeed scheduler metrics

### LINKED ISSUE #110420: rangefeed: add metrics to rangefeed scheduler
Current scheduler doesn't expose any metrics. This could be limiting as we don't have a visibility on how long rangefeed processors could wait in the queue before their work is scheduled for execution.

We need to at least have a wait time histogram and a pending queue size.

This visibility is critical to measuring how feeds could affect eachother and how timely system events are delivered over rangefeeds.

Jira issue: CRDB-31419

Epic CRDB-26372

### PR DESCRIPTION
**Pull request by aliher1911**
This commit adds two rangefeed scheduler metrics to assist performance debugging.

Latency histogram which is showing how long events spent waiting in queue before they are scheduled for processing:
- `kv.rangefeed.scheduler.normal.latency`
- `kv.rangefeed.scheduler.system.latency`

Queue size which is showing how many ranges are pending event to be processed:
- `kv.rangefeed.scheduler.normal.queue_size`
- `kv.rangefeed.scheduler.system.queue_size`

Fixes: #110420

Release note: None

### CODE REVIEW

**Review by miretskiy — APPROVED** [review]

File: pkg/kv/kvserver/rangefeed/scheduler.go [hunk 1]
```diff
@@ -323,33 +349,44 @@ func (ss *schedulerShard) unregister(id int64) {
...
 		return false
 	}
 	if pending == 0 {
+		// Only keep timestamp on entries every HistogramFrequency times.
+		if ss.nextLatencyCheck <= 1 {
+			ss.nextLatencyCheck = ss.histogramFrequency
+		} else {
+			entry.startTime = 0
+		}
+		ss.nextLatencyCheck--
```
[the remark this answers is not in the document]
Comment (aliher1911) [reply pkg/kv/kvserver/rangefeed/scheduler.go:382]: I think scheduler latency gives highest latency of the batch which is what we want internally. For events it makes more sense to measure end to end from the point where we enqueue logical ops to the point where we pass it to grpc sink. But that doesn't give us enough visibility on where we spend time.

Once we switch to buffered registrations, we will have work queue of multiple ranges combined which will further hide the sources of delays. I think both of those values are important for different reasons.
Currently checkpoints give us a processing latency if we need that I think.

I also measured batches and in normal case we don't have more than a 1 or 2 objects per batch so we really measuring every event here.

[the remark this answers is not in the document]
Comment (erikgrinaker) [reply pkg/kv/kvserver/rangefeed/scheduler.go:382]: Ok, let's just do the scheduler latency here for now, we can do event latency later. But in that case, we should just measure the actual scheduler latency, i.e. we shouldn't carry over `startTime` when we re-schedule work -- doing so results in this odd quantity where we measure rangefeed event latency but counting it by scheduling events, which results in neither scheduling latency nor event latency.

File: pkg/kv/kvserver/rangefeed/scheduler.go [hunk 2]
```diff
@@ -323,33 +349,44 @@ func (ss *schedulerShard) unregister(id int64) {
 // enqueue enqueues a single event for a given processor in this shard, and wakes
 // up a worker to process it. The caller must not hold the shard lock.
 func (ss *schedulerShard) enqueue(id int64, evt processorEventType) {
+	// We get time outside of lock to get more realistic delay in case there's a
+	// scheduler contention.
+	now := timeutil.Now().UnixNano()
```
[the remark this answers is not in the document]
Comment (aliher1911) [reply pkg/kv/kvserver/rangefeed/scheduler.go]: Changed to only get time when needed and on batch enqueues. Should be less wasteful now!

File: pkg/kv/kvserver/rangefeed/scheduler.go [hunk 3]
```diff
@@ -359,18 +401,24 @@ func (ss *schedulerShard) enqueueN(ids []int64, evt processorEventType) int {
 		return 0
 	}
 
+	// For bulk insertions we ignore sampling frequency and just use one sample
+	// per chunk for simplicity.
+	now := timeutil.Now().UnixNano()
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/rangefeed/scheduler.go:406]: Why not just use the actual sampling frequency, i.e. `i%histogramFrequency`? It's no more code than what's already here.

**Review by erikgrinaker — APPROVED** [review]

**Comment by aliher1911** [comment]
bors r=erikgrinaker,miretskiy

**Pull request merged**